File: 0028-find-the-index-of-the-first-occurrence-in-a-string/0028-find-the-index-of-the-first-occurrence-in-a-string.cpp
class Solution {
public:
    int strStr(string haystack, string needle) {
        // using two for loops 
        int hLen = haystack.size();
        int nLen = needle.size();
        
        // outer loop select all possible starting points in haystack
        for(int i = 0; i <= hLen - nLen; i++) {
            // will loop through needle
            int j = 0;
            for(; j < nLen; j++) {
                if(haystack[i + j] != needle[j]) {
                    break;
                }
            }
            if(j == nLen) return i;
        }
        
        return -1;
    }
};

// tc -> O(n*m)
// sc -> O(1)

File: 0028-find-the-index-of-the-first-occurrence-in-a-string/0028-find-the-index-of-the-first-occurrence-in-a-string.js
/**
 * @param {string} haystack
 * @param {string} needle
 * @return {number}
 */
var strStr = function(haystack, needle) {
    if(needle === "") return -1;
    
    const n = haystack.length;
    const m = needle.length;
    
    for(let i = 0; i <= n - m; i++) {
        for(let j = 0; j < m; j++) {
            if(haystack[i + j] !== needle[j]) break;
            
            if(j === m - 1) return i;
        }
    }
    
    return -1;
};